File: Leetcode/Medium/143. Reorder List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    void reorderList(ListNode *head)
    {
        ListNode *prev = nullptr;
        ListNode *slow = head;
        ListNode *fast = head;
        while (fast != nullptr && fast->next != nullptr)
        {
            fast = fast->next->next;
            prev = slow;
            slow = slow->next;
        }
        // head has odd number of nodes
        if (fast != nullptr)
        {
            prev = slow;
            slow = slow->next;
        }
        prev->next = nullptr;
        prev = nullptr;
        ListNode *next = nullptr;
        while (slow)
        {
            next = slow->next;
            slow->next = prev;
            prev = slow;
            slow = next;
        }
        // Merge 2 lists
        ListNode *head1 = head;
        ListNode *head2 = prev;
        ListNode *dummy = new ListNode(0);
        ListNode *curr = dummy;
        int cnt = 0;
        while (head1 != nullptr || head2 != nullptr)
        {
            if (head1 == nullptr)
            {
                curr->next = head2;
                break;
            }
            if (head2 == nullptr)
            {
                curr->next = head1;
                break;
            }
            if (cnt % 2 == 0)
            {
                curr->next = head1;
                head1 = head1->next;
            }
            else
            {
                curr->next = head2;
                head2 = head2->next;
            }
            curr = curr->next;
            cnt++;
        }
        head = dummy->next;
    }
};

File: Leetcode/Medium/143. Reorder List/143. Reorder List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.

        0 -> 1 -> 2 -> 3

        fast None (null Pointer)

        0 -> 3 -> 1 -> 2

        """
        if not head or not head.next: return head
        fast, slow = head, head 
        prev = None
        while fast and fast.next:
            prev = slow
            slow = slow.next
            fast = fast.next.next
        if fast:
            prev = slow
            slow = slow.next
        prev.next = None
        prev = None
        while slow:
            tmp = slow.next
            slow.next = prev
            prev = slow
            slow = tmp
        head1 = head
        head2 = prev
        dummy = ListNode()
        curr = dummy
        cnt = 0
        while head1 or head2:
            if not head2:
                curr.next = head1
                break
            curr.next = head1
            head1 = head1.next
            curr = curr.next
            curr.next = head2
            head2 = head2.next
            curr = curr.next
        head = dummy.next
        
        